### DISCUSSION TITLE
Disconnect is not fired on root namespace

### DISCUSSION THREAD

**Discussion by schen96** [miguelgrinberg/Flask-SocketIO#1805] [Q&A] [answered]
Hello, 

To preface i'm not sure if this is a bug or intended behaviour but basically from my understanding when socket-io client creates any N number of namespaces, they all share the same WS connection. 

In my application, I have two events

```
@socket_io.on("connect", namespace="/notify")
@requires_auth(socket_io=True)
def connected():
    # keep track of connection counts
    print("CONNECTED!!")


@socket_io.on("connect")
def root_connect():
    print("Root connection established")
```

When the user is connected, "Root connection established" is logged to console. If the user is authenticated, "Root connection is established" followed by "CONNECTED!!" is logged on to console. What i take from this is that the connect on root namespace represents the WS connection in which /notify namespace uses. Maybe I'm understanding this wrong. 

Now I test the "disconnect" event. My expectations are that "disconnect" event on root namespace will always fire.

```
@socket_io.on("disconnect", namespace="/notify")
def root_disconnect():
    print("disconnected namespace")


@socket_io.on("disconnect")
def root_disconnect():
    print("disconnected root")
```

However, 'disconnected root' is never logged, regardless if user is authenticated or not (although i dont think its important detail here). Only 'disconnected namespace' is logged. Is this a bug? If 'Root connection established' is logged shouldn't 'disconnected root' also be logged? 

The reason i ask is that I want to keep track of the number of WS connections. If there are say 10 namespaces, I don't want to log 10 connections if they are all sharing the same WS connection. I thought I could keep track of WS connection by looking at "connected" event on root namespace, but the "disconnect" namespace behaviour makes me a bit confused.

Anyone have any thoughts on this? Thank `you!`

**Comment by miguelgrinberg** [comment, marked as answer]
There is no concept of a "root" namespace, this is actually an outdated thing that existed in older versions of Socket.IO. Current versions treat all namespaces equal, and you can connect and disconnect from each namespace independently.

**Reply by schen96** [reply]
Hi i'm using Flask-Socketio 4.3. Maybe "root" is not the correct term but the default namespace "/" always connects. This is evident from testing and you mentioned this in your documentation on migrating from Flask-SocketIO 4 -> 5

"The default namespace '/' is not automatically connected anymore, and is now treated in the same way as
other namespaces."

If the default namespace connects in versions < 5, shouldn't it follow that disconnect event on default namespace should also fire when client disconnects? 

Thank you

**Reply by miguelgrinberg** [reply]
On the old releases, yes. When the client disconnects from the default namespace you should get the disconnect event fired. You did not mention anything about how you are disconnecting. My guess is that you are disconnecting from your non-default namespace, while keeping the default namespace connected.

In any case, this behavior of namespace was extremely confusing to people. The current model is more reasonable, in my opinion.

**Reply by schen96** [reply]
I would like to find the number of WS connections. Anyway to do this? I see people incrementing a global counter on connect / disconnect events, but since namespaces share the same WS connection this seems wrong. If you have two namespaces /A /B then this mechanism would count two connections, even though they are both sharing one connection

**Reply by schen96** [reply]
Ive tried closing browser tab and closing browser itself on the client. I feel like that shoudl force a disconnect no?

**Reply by miguelgrinberg** [reply]
> I would like to find the number of WS connections. Anyway to do this?

This is better done in the client. Look in the network tab for your active WS connections.

> Ive tried closing browser tab and closing browser itself on the client. I feel like that shoudl force a disconnect no?

Yes, it should. I don't have enough context from you to tell you more. It could be a bug, either one that was fixed at some point, or maybe one that was never reported, so it was never addressed. Note that I am not maintaining those releases anymore, only the current 5.x line is open.

**Reply by schen96** [reply]
> Yes, it should. I don't have enough context from you to tell you more. It could be a bug, either one that was fixed at some point, or maybe one that was never reported, so it was never addressed. Note that I am not maintaining those releases anymore, only the current 5.x line is open.

Okay I thank you I appreciate your response. I will check with other versions. Love the extension.
